File: day07/part1.go
package main

import (
"bufio" 
"fmt" 
"os"
"strings"
)

type Dependency struct {
	from, to byte
}

func readFile(filename string) []Dependency {
	file, _ := os.Open(filename)
	defer file.Close()

	var deps []Dependency
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		from, end := 5, 36
		line := scanner.Text()
		deps = append(deps, Dependency{line[from] - 'A', line[end] - 'A'})
	}

	return deps;
}

func createTasks(deps []Dependency) [26]string {
	var tasks [26]string
	for i := 0; i < 26; i++ {
		tasks[i] = "?"
	}

	for _, dep := range deps {
		newDep := string('A' + dep.from)

		if tasks[dep.from] == "?" {
			tasks[dep.from] = ""
		}
		if tasks[dep.to] == "?" {
			tasks[dep.to] = ""
		}

		tasks[dep.to] = tasks[dep.to] + newDep
	}

	return tasks
}

func buildSequence(tasks [26]string) string {
	sequence := ""
	loop := true
	for loop {
		loop = false
		for i := 0; i < 26 && !loop; i++ {
			if tasks[i] == "" {
				elem := string('A' + i)
				sequence += elem
				loop = true
				tasks[i] = "?"

				for j := 0; j < 26; j++ {
					tasks[j] = strings.Replace(tasks[j], elem, "", -1)
				}
			}
		}
	}
	return sequence
}

func main() {
	fmt.Println(buildSequence(createTasks(readFile("input"))))
}

File: day03/part1.py
import re

def readfile(filename):
    with open(filename) as file:
        return [line.rstrip('\n') for line in file]

def extract(line):
    return map(int, re.match(r'#(\d+)\s*@\s*(\d+),(\d+):\s*(\d+)x(\d+)', line).groups())

def main():
    n = 1000
    fabric = [[0 for _ in range(n)] for _ in range(n)]

    for line in readfile('input'):
        _, x, y, w, h = extract(line)
        for i in range(w):
            for j in range(h):
                fabric[x + i][y + j] += 1

    count = 0
    for i in range(n):
        for j in range(n):
            if fabric[i][j] > 1:
                count += 1

    print(count)

main()
